# Probe Sequences Contents

<b>O mykiss probes:</b>  

(1) Omy_391_probe_seqs.csv: <b>USE THIS ONE</b> probe sequences for the SFGL O mykiss panel as of December 2020, corrected alleles for the Omy05 residency markers. Omy_RAD47080-54 (which is a duplicate to Omy_RAD15709-53) is removed. Species marker Oki120255-113 is added.  
(2) Omy_390_probe_seqs.csv: probe sequences for the SFGL O mykiss panel as of December 2020, corrected alleles for the Omy05 residency markers. Omy_RAD47080-54 (which is a duplicate to Omy_RAD15709-53) is removed.   
(3) Omy_390_probe_seqs_uncorrected.csv: probe sequences for the SFGL O mykiss panel as of December 2020, does not contain corrected alleles for the Omy05 residency markers  

<b>O. tshawytscha:  </b> 

(1) Ots356_probeseqs.csv: This is the updated probe sequence file for that works with 75bp sequencing (referred to as OtsGTseqV6.1 at CRITFC). It includes 351 markers from CRITFC as well 5 "RoSA" markers.  
(2) Ots_335_probe_seqs.csv: must up to date version of the non-75bp compliant Ots probe sequence, but includes a duplicate  
(3) Ots_334_probe_seqs_removed_Ots37124-12279142.csv: Includes most recent Ots28 markers, but excludes a duplicate marker identified by WDFW  
(4) Ots_335_probe_seqs.txt: Does not include latest Ots28 markers  
(5) **Ots352: USE THIS ONE** david's best guess as to what is actually in the panel (exclude read hogs, add back the excluded rosa marker (note that I included the reverse primer as FWD primer and changed the SNP codes and probes to their reverse complements), include "miscellaneous plate")  

<b>O. keta:  </b>  

(1) Oke_GTseq350_ProbeSeqs.csv: Full Oke probe sequence file

<b> T. alalunga (Albacore): </b>  

(1) Tal289_probe_seqs.csv: Updated (Feb 2022) to include Nate Campbell's optimized allele correction values and probe sequences. 

<b>S. confluentus (Bull Trout): </b>

(1) Sco_GTseq266_263autosomes_ProbeSeqs.csv: Sco and Sfo probe sequence file
